### PR TITLE
Add QgsMatrix4x4

### PR DESCRIPTION
**Pull request by wonder-sk**
This adds 4x4 matrix with double precision values for high precision transforms. Just like QgsVector3D is similar to QVector3D but uses doubles instead of floats, this is a similar approach with QgsMatrix4x4 vs QMatrix4x4.

Compared to the Qt's QMatrix4x4, the API has been reduced to the absolute minimum to only provide functionality that we will need (create matrix, get matrix data, multiply with other matrix, multiply with vector).

### CODE REVIEW

File: src/core/qgsmatrix4x4.cpp [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+  if ( m[2][1] != 0.0f || m[2][2] != 1.0f || m[2][3] != 0.0f )
+    return false;
+  if ( m[3][0] != 0.0f || m[3][1] != 0.0f || m[3][2] != 0.0f )
+    return false;
+  return ( m[3][3] == 1.0f );
+}
+
+void QgsMatrix4x4::setToIdentity()
+{
+  m[0][0] = 1.0f;
```
Comment (lbartoletti) [inline src/core/qgsmatrix4x4.cpp:76]:
```suggestion
  m[0][0] = 1.0;
```

'f' suffix is for 'float', 'l' for 'long double', no suffix for double: https://en.cppreference.com/w/cpp/language/floating_literal

File: src/core/qgsmatrix4x4.cpp [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+      matrix.m[3][3];
+  if ( w == 1.0f )
+    return QgsVector3D( x, y, z );
+  else
+    return QgsVector3D( x / w, y / w, z / w );
+}
+
+bool QgsMatrix4x4::isIdentity() const
+{
+  if ( m[0][0] != 1.0f || m[0][1] != 0.0f || m[0][2] != 0.0f )
```
Comment (lbartoletti) [inline src/core/qgsmatrix4x4.cpp:61]:
```suggestion
  if ( m[0][0] != 1.0 || m[0][1] != 0.0 || m[0][2] != 0.0 )
```

Comment (lbartoletti) [reply src/core/qgsmatrix4x4.cpp:61]: same elsewhere

File: src/core/qgsmatrix4x4.h [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+ *
+ * It is similar to QMatrix4x4, but working with double precision values.
+ * Most of the time, doing transform using QMatrix4x4 is fine, however QgsMatrix4x4
+ * is useful in situations where single precision floats are not enough.
+ * For example, when using transform matrix where translation component has
+ * values in order of millions.
+ *
+ * \warning Only for internal use
+ *
+ * \note Not available in Python bindings
```
Comment (lbartoletti) [inline src/core/qgsmatrix4x4.h:37]: why? Later, I guess, it could be useful.

File: tests/src/core/testqgsmatrix4x4.cpp [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+
+//runs after all tests
+void TestQgsMatrix4x4::cleanupTestCase()
+{
+}
+
+void TestQgsMatrix4x4::testData()
+{
+  // Initialization - row-major order
+  QgsMatrix4x4 m1( 10, 20, 30, 40,
```
Comment (lbartoletti) [inline tests/src/core/testqgsmatrix4x4.cpp:54]:
```suggestion
  QgsMatrix4x4 m1( 10.0, 20.0, 30.0, 40.0,
```

or simply "x.", but avoid implicit conversions

File: tests/src/core/testqgsmatrix4x4.cpp [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+  QCOMPARE( mdata[14], 0. );
+  QCOMPARE( mdata[15], 1. );
+
+  QVERIFY( m.isIdentity() );
+}
+
+void TestQgsMatrix4x4::testVectorMultiply()
+{
+  double sa = 0.5, ca = sqrt( 3 ) / 2; // sin(30 deg) and cos(30 deg)
+  QgsMatrix4x4 m( ca, -sa, 0, 0,
```
Comment (lbartoletti) [inline tests/src/core/testqgsmatrix4x4.cpp:113]:
```suggestion
  QgsMatrix4x4 m( ca, -sa, 0.0, 0.0,
```

same

File: src/core/qgsmatrix4x4.h (hunk 3 above)
Comment (wonder-sk) [reply src/core/qgsmatrix4x4.h:37]: It is a low-level class, an implementation detail that is not suited for public API use. We can make it a part of Python API when there will be a cause for it...

**Comment by wonder-sk** [comment]
@lbartoletti thanks for the review. I think the float/int -> double conversion of literals is done automatically by the compiler, so there should not be any performance penalty if they are not double literals in the source code. I have updated the code nevertheless in [7bdc3df](https://github.com/qgis/QGIS/pull/53842/commits/7bdc3dfbd6f09bffd4d119b42a6f1e2406fb3f22)

File: src/core/qgsmatrix4x4.cpp [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+      matrix.m[3][3];
+  if ( w == 1.0f )
+    return QgsVector3D( x, y, z );
+  else
+    return QgsVector3D( x / w, y / w, z / w );
+}
+
+bool QgsMatrix4x4::isIdentity() const
+{
+  if ( m[0][0] != 1.0 || m[0][1] != 0.0 || m[0][2] != 0.0 )
```
Comment (nyalldawson) [inline src/core/qgsmatrix4x4.cpp:61]: Should we be using qgsDoubleNear here? How exact does the intended use case for "isIdentity" need to be?

File: src/core/qgsmatrix4x4.h (hunk 3 above)
Comment (nyalldawson) [reply src/core/qgsmatrix4x4.h:37]: FWIW, I'm currently leaning toward an "expose to python, but mark as non-stable API" for situations like this. My rationale is:

- it makes test development MUCH quicker (no recompilation necessary), which ultimately results in better test coverage
- it avoids the forced creation of additional .cpp tests , which all add non-negligible recompilation/linking costs
- there's nothing more frustrating when writing a plugin then bumping into a situation where something isn't exposed in PyQGIS, and you either need to re-write a bunch of existing cpp code into Python or are forced to expose the method in QGIS master, wait for release, and put an aggressive minimum QGIS version on the plugin users. While it's not particularly nice to have exposed classes marked as non-stable, I feel it's the lesser of two evils, and with a plugin development hat on I'd rather wear the risk of a non-stable api changing vs being forced to rewrite stuff from QGIS c++ in Python myself.

File: src/core/qgsmatrix4x4.cpp (hunk 6 above)
Comment (wonder-sk) [reply src/core/qgsmatrix4x4.cpp:61]: This is borrowed from Qt's implementation where they use exact comparison. I would stick to it, this behavior seems more correct, and best to keep the implementation close to Qt's anyway.

I think the intended use case for "isIdentity" is more like "has this been constructed by the default constructor"? I used it in the 3d tiles code to check if we need to apply transform and recalculate bounding volumes...

File: src/core/qgsmatrix4x4.h (hunk 3 above)
Comment (wonder-sk) [reply src/core/qgsmatrix4x4.h:37]: Ok, fair enough, I will expose it...

As for C++/Python tests, I agree with the point of more compilation to do, but on the other hand, debugging PyQGIS tests when they break is more pain :-)

Comment (wonder-sk) [reply src/core/qgsmatrix4x4.h]: Python support added in [2881df9](https://github.com/qgis/QGIS/pull/53842/commits/2881df99f50f472c4d399ec191ef9191e2dac2cd)

**Review by nyalldawson — APPROVED** [review]

**Pull request merged**